lib/graphql/query/serial_execution/field_resolution.rb

```diff
@@ -47,21 +53,35 @@ def get_finished_value(raw_value)
             end
           end
 
-          begin
-            GraphQL::Query::SerialExecution::ValueResolution.resolve(
-              parent_type,
-              field,
-              field.type,
-              raw_value,
-              @irep_nodes,
-              @field_ctx,
-            )
-          rescue GraphQL::InvalidNullError => err
-            if field.type.kind.non_null?
-              raise(err)
-            else
-              err.parent_error? || @query.context.errors.push(err)
-              nil
+          box_method = @query.boxed_value_method(raw_value)
+          if box_method
+            GraphQL::Execution::Boxed.new { raw_value.public_send(box_method) }.then { |val|
+              GraphQL::Query::SerialExecution::ValueResolution.resolve(
+                parent_type,
+                field,
+                field.type,
+                val,
+                @irep_nodes,
+                @field_ctx,
+              )
+            }
+          else
+            begin
+              GraphQL::Query::SerialExecution::ValueResolution.resolve(
+                parent_type,
+                field,
+                field.type,
+                raw_value,
+                @irep_nodes,
+                @field_ctx,
+              )
+            rescue GraphQL::InvalidNullError => err
```

Comment:
Oh, I misunderstood! Yes, I definitely plain to maintain all the existing behavior ... but propagating nulls is harder when the response is already half-built :S 

I'm exploring an approach where you keep the type info alongside the result, so that if a lazy value returns nil, you can backtrack to find a nullable field.